### PR TITLE
Feature/rap 33 interim request form

### PR DESCRIPTION
**Pull request by Luisella21**

### CODE REVIEW

File: chartofaccount/models.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
+from django.db import models
+
+
+class DepartmentalGroup(models.Model):
+    group_code = models.CharField("Group Code", primary_key=True, max_length=6)
+    group_name = models.CharField("Group Name", max_length=300)
+
+    @property
+    def full_name(self):
+        return "{self.group_code} - {self.group_name}"
```
Comment (SamDudley) [inline chartofaccount/models.py:10]: This is missing the `f` for an `f` string.

File: chartofaccount/models.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    directorate_code = models.CharField(
+        "Directorate Code", primary_key=True, max_length=6
+    )
+    directorate_name = models.CharField("Directorate Name", max_length=300)
+    group = models.ForeignKey(
+        DepartmentalGroup, on_delete=models.CASCADE, related_name="directorates"
+    )
+
+    def __str__(self):
+        return str(self.directorate_name)
```
Comment (SamDudley) [inline chartofaccount/models.py:29]: No need to call `str` as it will always be one.

File: chartofaccount/models.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+class DepartmentalGroup(models.Model):
+    group_code = models.CharField("Group Code", primary_key=True, max_length=6)
+    group_name = models.CharField("Group Name", max_length=300)
+
+    @property
+    def full_name(self):
+        return "{self.group_code} - {self.group_name}"
+
+    def __str__(self):
+        return str(self.group_name)
```
Comment (SamDudley) [inline chartofaccount/models.py:13]: No need to call `str` as it will always be one.

File: chartofaccount/models.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        primary_key=True,
+        max_length=50,
+    )
+    programme_description = models.CharField(
+        "Programme Name",
+        max_length=100,
+    )
+
+    def __str__(self):
+        return self.programme_code + " - " + self.programme_description
```
Comment (SamDudley) [inline chartofaccount/models.py:67]: You have used `f` strings in the other models and it should be consistent.

File: chartofaccount/models.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        "Project Code",
+        primary_key=True,
+        max_length=50,
+    )
+    project_description = models.CharField(
+        max_length=300, verbose_name="Project Description"
+    )
+
+    def __str__(self):
+        return self.project_code + " - " + self.project_description
```
Comment (SamDudley) [inline chartofaccount/models.py:85]: Convert to `f` string.

File: chartofaccount/tests.py [hunk 6]
```diff
@@ -0,0 +1,3 @@
+from django.test import TestCase
```
Comment (SamDudley) [inline chartofaccount/tests.py:1]: File can be removed.

File: chartofaccount/views.py [hunk 7]
```diff
@@ -0,0 +1,3 @@
+from django.shortcuts import render
```
Comment (SamDudley) [inline chartofaccount/views.py:1]: File can be removed.

File: main/templates/main/statement_of_work.html [hunk 8]
```diff
@@ -4,6 +4,9 @@
 
 {% block content %}
 <h2 class="govuk-heading-m">{{ title }}</h2>
+
+    <script src="https://unpkg.com/htmx.org@1.6.0"></script>
```
Comment (SamDudley) [inline main/templates/main/statement_of_work.html:8]: Switch to a script tag which includes an SRI for added security.

image

File: main/templates/main/interim_request.html [hunk 9]
```diff
@@ -0,0 +1,18 @@
+{% extends 'main/base.html' %}
+
+{% block title %}{{ title }}{% endblock %}
+
+{% block content %}
+<h2 class="govuk-heading-m">{{ title }}</h2>
+
+<script src="https://unpkg.com/htmx.org@1.6.0"></script>
```
Comment (SamDudley) [inline main/templates/main/interim_request.html:8]: Switch to a script tag with SRI (see the other comment below).

File: Makefile [hunk 10]
```diff
@@ -26,6 +26,15 @@ groups:
 
 test-data: groups	
 	$(web) python manage.py loaddata test-users.json
+	$(web) python manage.py loaddata chartofaccount.json
```
Comment (SamDudley) [inline Makefile:29]: If `chartofaccount.json` is test data only can we rename it to `test-chartofaccount.json` so that it is obvious to all?

**Review by SamDudley — CHANGES_REQUESTED** [review]

File: main/forms/interim_request_form.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                }
+            ),
+        }
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.fields["resourcing_request"].disabled = True
+
+        if "group" in self.data:
```
Comment (SamDudley) [inline main/forms/interim_request_form.py:38]: Add comments explaining the what each branch is at a high level.

File: main/forms/interim_request_form.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+            try:
```
Comment (SamDudley) [inline main/forms/interim_request_form.py:39]: Try refactor into a function within the utils.py file. The function will take the form (self) and perform the field updates for the related selects.

File: main/models.py [hunk 13]
```diff
@@ -218,14 +226,43 @@ class StatementOfWork(models.Model):
...
     role = models.CharField(max_length=255)
-    cost_centre_code = models.CharField(max_length=6)
-    programme_code = models.CharField(max_length=4)
-    project_code = models.CharField(max_length=6, blank=True, null=True)
+    project_description = models.TextField()
+
+    group = models.ForeignKey(
+        DepartmentalGroup,
+        on_delete=models.CASCADE,
+        related_name="statement_of_work_groups",
```
Comment (SamDudley) [inline main/models.py:235]: Change to `"+"`.

File: main/models.py [hunk 14]
```diff
@@ -310,7 +351,86 @@ class InterimRequest(models.Model):
...
+    )
+    part_b_main_reason = models.TextField(
+        verbose_name="",
+        help_text="What are the main reasons why this role has not been filled by a substantive Civil Servant. Please detail the strategic workforce plan for this role after the assignment end date:",
+    )
+
+    group = models.ForeignKey(
+        DepartmentalGroup,
+        on_delete=models.CASCADE,
+        related_name="interim_request_groups",
```
Comment (SamDudley) [inline main/models.py:417]: Same `"+"`.

**Review by SamDudley — CHANGES_REQUESTED** [review]

File: chartofaccount/tests.py (hunk 6 above)
Comment (Luisella21) [reply chartofaccount/tests.py]: Done.

File: Makefile (hunk 10 above)
Comment (Luisella21) [reply Makefile]: Done.

File: main/forms/interim_request_form.py (hunk 12 above)
Comment (Luisella21) [reply main/forms/interim_request_form.py]: Done.

File: main/models.py (hunk 13 above)
Comment (Luisella21) [reply main/models.py]: Done.

**Review by SamDudley — APPROVED** [review]
LGTM!

**Pull request merged**